File: python/src/threads/loop/questions.py
"""ask_user in the loop (spec/schema/README.md, "Questions and remembered rules"): a valid
question parks the branch for 24 hours; one the question rules refuse is an error result. An
unanswered question past its expiry is closed with "no answer" by the next run on the branch."""

from typing import TYPE_CHECKING, Final

from threads._generated.tools_v1 import AskUserInput
from threads.log import ParkedEvent
from threads.log.ask_user import ask_problem
from threads.loop.drafts import draft
from threads.loop.history import CallState
from threads.loop.results import As, result_draft
from threads.loop.runtime import Halt, Parked, Runtime, lost
from threads.result import Err
from threads.store import Draft
from threads.store.lines import uuid7

if TYPE_CHECKING:
    from pydantic import JsonValue

TTL_MS: Final = 86_400_000
NO_ANSWER: Final = "no answer"


async def ask(rt: Runtime, state: CallState) -> Halt | None:
    """Parks on {input, call_id}, or closes the call when its input breaks the question rules."""
    call_id = state.call.data.call_id
    why = ask_problem(AskUserInput.model_validate(dict(state.call.data.input)))
    if why is not None:
        result = await result_draft(rt, call_id, f"invalid input: {why}", As("not_executed", True))
        done = await rt.append(result)
        return lost(done.error) if isinstance(done, Err) else None
    data: dict[str, JsonValue] = {
        "address": {"kind": "input", "id": call_id},
        "reason": "awaiting_input",
        "expires_at": rt.clock() + TTL_MS,
    }
    done = await rt.append(draft("parked", data))
    if isinstance(done, Err):
        return lost(done.error)
    return Parked("awaiting_input", tuple(rt.fold.parked))


async def expire(rt: Runtime) -> Halt | None:
    """Each open question whose park has expired: the error result "no answer", then resumed."""
    now = rt.clock()
    for event in list(rt.events):
        if not isinstance(event, ParkedEvent) or event.data.address not in rt.fold.parked:
            continue
        address, at = event.data.address, event.data.expires_at
        if address.kind != "input" or not isinstance(at, int) or at > now:
            continue
        cause = uuid7(now)
        data: dict[str, JsonValue] = {
            "call_id": address.id,
            "is_error": True,
            "origin": "not_executed",
            "completeness": "complete",
            "preview": NO_ANSWER,
        }
        resumed: dict[str, JsonValue] = {
            "address": {"kind": "input", "id": address.id},
            "cause_event_id": cause,
        }
        done = await rt.append(
            Draft("tool_result", data, {"kind": "host"}, True, cause), draft("resumed", resumed)
        )
        if isinstance(done, Err):
            return lost(done.error)
    return None

File: typescript/packages/core/src/loop/questions.ts
import type { EventOf, Fold, ParkAddress } from "../fold/state";
import type { KnownEvent } from "../log";
import { AskUserInput } from "../tools/agent-inputs";
import { askProblem } from "../tools/ask-user";
import { draft, HOST, TOOL } from "./drafts";
import type { Session } from "./session";
import type { Halt } from "./types";

// ask_user (spec/schema/README.md, "Questions and remembered rules"): a valid question parks the
// branch on {input, call_id} for a day; once that has passed with no answer, the next run on the
// branch records "no answer" and goes on. A host's tick only has to start that run.

const DAY_MS = 86_400_000;

export function askUser(
  s: Session,
  call: EventOf<"tool_call">,
): Promise<Halt | undefined> {
  const { call_id } = call.data;
  // Arguments already parsed before authorization; the question rules are rule 48.
  const problem = askProblem(AskUserInput.parse(call.data.input));
  if (problem !== undefined)
    return s.append(
      draft.toolResult(
        {
          call_id,
          is_error: true,
          origin: "not_executed",
          preview: `invalid input: ${problem}`,
        },
        TOOL,
      ),
    );
  return s.append(
    draft.parked({
      address: { kind: "input", id: call_id },
      reason: "awaiting_input",
      expires_at: s.now() + DAY_MS,
    }),
  );
}

/** The open questions whose expires_at has passed, in the order they were asked. */
export function dueQuestions(
  events: readonly KnownEvent[],
  fold: Fold,
  now: number,
): readonly ParkAddress[] {
  return fold.parked.filter((address) => {
    if (address.kind !== "input") return false;
    const asked = events.findLast(
      (e) =>
        e.type === "parked" &&
        e.data.address.kind === "input" &&
        e.data.address.id === address.id,
    );
    const expires =
      asked?.type === "parked" ? asked.data.expires_at : undefined;
    return expires !== undefined && expires <= now;
  });
}

/** Closes each due question with the error result "no answer", then resumes it. */
export async function expireQuestions(s: Session): Promise<Halt | undefined> {
  for (const address of dueQuestions(s.events, s.fold, s.now())) {
    const closed = await s.append(
      draft.toolResult(
        {
          call_id: address.id,
          is_error: true,
          origin: "not_executed",
          preview: "no answer",
        },
        HOST,
      ),
    );
    if (closed !== undefined) return closed;
    const result = s.events.at(-1);
    if (result === undefined) throw new Error("the result was just appended");
    const stopped = await s.append({
      type: "resumed",
      type_version: 1,
      critical: true,
      actor: HOST,
      data: { address, cause_event_id: result.event_id },
    });
    if (stopped !== undefined) return stopped;
  }
  return undefined;
}
